src/openms/include/OpenMS/ANALYSIS/ID/IDBoostGraph.h

```diff
@@ -0,0 +1,278 @@
+    void clusterIndistProteinsAndPeptides();
+
+    /// Annotate indistinguishable proteins by adding the groups to the underlying
+    /// ProteinIdentification::ProteinGroups object. This has no effect on the graph itself.
+    /// @param addSingletons if you want to annotate groups with just one protein entry
+    void annotateIndistProteins(bool addSingletons = true) const;
+
+
+    /// Visits nodes in the boost graph (ptrs to an ID Object) and depending on their type creates a label
+    class LabelVisitor:
+        public boost::static_visitor<OpenMS::String>
+    {
+    public:
+
+      OpenMS::String operator()(const PeptideHit* pep) const
+      {
+        return pep->getSequence().toUnmodifiedString();
+      }
+
+      OpenMS::String operator()(const ProteinHit* prot) const
+      {
+        return prot->getAccession();
+      }
+
+      OpenMS::String operator()(const ProteinGroup* /*protgrp*/) const
+      {
+        return String("PG");
+      }
+
+      OpenMS::String operator()(const PeptideCluster* /*pc*/) const
+      {
+        return String("PepClust");
+      }
+
+    };
+
+    /// Visits nodes in the boost graph (ptrs to an ID Object) and depending on their type prints the address. For debug
+    class PrintAddressVisitor:
+        public boost::static_visitor<>
+    {
+    public:
+
+      void operator()(PeptideHit* pep) const
+      {
+        std::cout << pep->getSequence().toUnmodifiedString() << ": " << pep << std::endl;
+      }
+
+      void operator()(ProteinHit* prot) const
+      {
+        std::cout << prot->getAccession() << ": " << prot << std::endl;
+      }
+
+      void operator()(const ProteinGroup* protgrp) const
+      {
+        std::cout << "PG" << std::endl;
+      }
+
+      void operator()(const PeptideCluster* /*pc*/) const
+      {
+        std::cout << "PepClust" << std::endl;
+      }
+
+    };
+
+    /// Visits nodes in the boost graph (ptrs to an ID Object) and depending on their type sets the posterior
+    /// Don't forget to set higherScoreBetter and score names in the parent ID objects.
+    class SetPosteriorVisitor:
+        public boost::static_visitor<>
+    {
+    public:
+
+      void operator()(PeptideHit* pep, double posterior) const
+      {
+        pep->setScore(posterior);
+      }
+
+      void operator()(ProteinHit* prot, double posterior) const
+      {
+        std::cout << "set score " << posterior << " for " << prot->getAccession() << std::endl;
+        prot->setScore(posterior);
+      }
+
+      void operator()(ProteinGroup* /*protgrp*/, double /*posterior*/) const
+      {
+        // do nothing
+        //protgrp->probability = posterior;
+      }
+
+      void operator()(const PeptideCluster* /*pc*/, double /*posterior*/) const
+      {
+        // do nothing
+      }
+
+    };
+
+    /// Visits nodes in the boost graph (ptrs to an ID Object) and depending on their type creates a random
+    /// variable or "Dependency" to add into the InferenceGraph
```

Comment:
clean up?

Reply:
Yup. Needs some cleanup. Agree. I can do it in this pr